### PR TITLE
nixos/pixiecore: init

### PR DESCRIPTION
**Pull request by bbigras**
###### Motivation for this change

This should only be merged after https://github.com/NixOS/nixpkgs/pull/83373 . You can use that PR to review the package part. This is one is for the module.

This is a rebase of #62113 (the module part).

I also added support for the "api" mode and added openFirewall and extraArguments options.

I tried to address all the reviews from the old abandoned PR.

The module was requested in #49740

@Lassulus @aanderse @Ekleog

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/pixiecore/default.nix [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  };
+
+  modSha256 = "1waqaglm6f9zy5296z309ppkck2vmydhk9gjnxrgzmhqld5lcq4f";
+  subPackages = [ "cmd/pixiecore" ];
+
+  meta = {
+    description = "A tool to manage network booting of machines";
+    homepage = "https://github.com/danderson/netboot/tree/master/pixiecore";
+    license =  stdenv.lib.licenses.asl20;
+    maintainers = with stdenv.lib.maintainers; [ bbigras ];
```
Comment (danderson) [inline pkgs/tools/networking/pixiecore/default.nix]: Hi! Pixiecore author here :)

You can add me as a maintainer as well (`danderson`, already in maintainers list), to increase bus factor.

File: nixos/modules/services/networking/pixiecore.nix [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      statusPort = mkOption {
+        type = types.port;
+        default = 80;
+        description = "HTTP port for status information (can be the same as --port)";
+      };
+
+      apiServer = mkOption {
+        type = types.str;
+        example = "127.0.0.1:8080";
+        description = "IPv4 address with port of the API. Ignored unless mode is set to 'api'";
```
Comment (danderson) [inline nixos/modules/services/networking/pixiecore.nix]: This can be a host:port too, it's "whatever Go accepts when dialing an HTTP server".

File: nixos/modules/services/networking/pixiecore.nix [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    };
+
+    systemd.services.pixiecore = {
+      description = "Pixiecore server";
+      after = [ "network.target"];
+      wants = [ "network.target"];
+      wantedBy = [ "multi-user.target"];
+      serviceConfig = {
+        Type="simple";
+        PIDFile="/run/pixiecore.pid";
```
Comment (danderson) [inline nixos/modules/services/networking/pixiecore.nix]: I don't think you need this. Pixiecore doesn't write out a PID file, so this file will never be populated, and will just confuse systemd and users.

File: nixos/modules/services/networking/pixiecore.nix [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        DynamicUser="yes";
```
Comment (danderson) [inline nixos/modules/services/networking/pixiecore.nix]: DynamicUser might make it tricky for Pixiecore to read kernel/initrd/etc. files from whatever location the user pointed at above. They'd have to make the files world-readable, since there's no predictable identity to grant access to. Maybe preferable to create a stable system user?

**Review by danderson — APPROVED** [review]
PR LGTM. Always happy to see Pixiecore in the wild :)

File: nixos/modules/services/networking/pixiecore.nix [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  cfg = config.services.pixiecore;
+in
+{
+  options = {
+    services.pixiecore = {
+      enable = mkEnableOption "Pixiecore";
+
+      openFirewall = mkOption {
+        type = types.bool;
+        default = true;
```
Comment (samueldr) [inline nixos/modules/services/networking/pixiecore.nix]:
```suggestion
        default = false;
```

Except for ssh, no port should be opened by enabling a service. (If another module does, it should be fixed.)

Comment (danderson) [inline nixos/modules/services/networking/pixiecore.nix]: (chanelling IRC discussion) Default here should be false. Except SSH, NixOS doesn't openFirewall by default, even if the module is enabled.

File: nixos/modules/services/networking/pixiecore.nix [hunk 6]
```diff
@@ -0,0 +1,112 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.pixiecore;
+in
+{
```
Comment (danderson) [inline nixos/modules/services/networking/pixiecore.nix]: Add a meta here, with at least maintainers.

(per my comment on the derivation, you're welcome to put me as backup maintainer here too)

**Comment by bbigras** [comment]
Thank you very much @danderson and @samueldr . I made all the changes.

**Comment by Mic92** [comment]
It would be also great if you could update the wiki after that.

File: nixos/modules/services/networking/pixiecore.nix [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+  options = {
+    services.pixiecore = {
+      enable = mkEnableOption "Pixiecore";
+
+      openFirewall = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Open ports in the firewall for Pixiecore.
```
Comment (Profpatsch) [inline nixos/modules/services/networking/pixiecore.nix]: I’d list the ports here.

File: nixos/modules/services/networking/pixiecore.nix [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+      };
+
+      dhcpNoBind = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Handle DHCP traffic without binding to the DHCP server port";
+      };
+
+      kernel = mkOption {
+        type = types.str;
```
Comment (Profpatsch) [inline nixos/modules/services/networking/pixiecore.nix]: This should be `str` or `path` so that you can give a derivation directly and don’t have to use `toString`

File: nixos/modules/services/networking/pixiecore.nix [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        default = "";
+        description = "Kernel path. Ignored unless mode is set to 'boot'";
+      };
+
+      initrd = mkOption {
+        type = types.str;
```
Comment (Profpatsch) [inline nixos/modules/services/networking/pixiecore.nix]: This should be `str` or `path` so that you can give a derivation directly and don’t have to use `toString`

File: nixos/modules/services/networking/pixiecore.nix [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+      serviceConfig = {
+        User = "pixiecore";
+        Restart = "always";
+        AmbientCapabilities = "cap_net_bind_service";
+        ExecStart = 
+          let
+            argString = if cfg.mode == "boot" then "boot ${cfg.kernel} ${cfg.initrd} --cmdline='${cfg.cmdLine}'"
+            else "api ${cfg.apiServer}";
+          in
+            "${pkgs.pixiecore}/bin/pixiecore ${argString} ${optionalString cfg.debug "--debug"} ${optionalString cfg.dhcpNoBind "--dhcp-no-bind"} --listen-addr '${cfg.listen}' --port ${toString cfg.port} --status-port ${toString cfg.statusPort} ${escapeShellArgs cfg.extraArguments}";
```
Comment (Profpatsch) [inline nixos/modules/services/networking/pixiecore.nix]:
```suggestion
            ''
              ${pkgs.pixiecore}/bin/pixiecore \
                ${argString} \
                ${optionalString cfg.debug "--debug"} \
                ${optionalString cfg.dhcpNoBind "--dhcp-no-bind"} \
                --listen-addr ${lib.escapeShellArg cfg.listen} \
                --port ${toString cfg.port} \
                --status-port ${toString cfg.statusPort} \
                ${escapeShellArgs cfg.extraArguments}
              '';
```

Comment (Profpatsch) [reply nixos/modules/services/networking/pixiecore.nix]: Linebreaks & better escaping of some commands

File: nixos/modules/services/networking/pixiecore.nix [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+      wants = [ "network.target"];
+      wantedBy = [ "multi-user.target"];
+      serviceConfig = {
+        User = "pixiecore";
+        Restart = "always";
+        AmbientCapabilities = "cap_net_bind_service";
+        ExecStart = 
+          let
+            argString = if cfg.mode == "boot" then "boot ${cfg.kernel} ${cfg.initrd} --cmdline='${cfg.cmdLine}'"
+            else "api ${cfg.apiServer}";
```
Comment (Profpatsch) [inline nixos/modules/services/networking/pixiecore.nix]:
```suggestion
```suggestion
            argString = 
              if cfg.mode == "boot"
              then [ "boot" cfg.kernel cfg.initrd "--cmdline=${lib.escapeArgString cfg.cmdLine}" ]
```
            else [ "api" cfg.apiServer ];
```

and then use `lib.escapeShellArgs` below

**Comment by bbigras** [comment]
Thanks for the review @Profpatsch . Is it OK like that or should I use `<literal>` around `port` and `statusPort`?

https://github.com/NixOS/nixpkgs/blob/3ae89bc8a19ac92f9b7e772c2da1cd7d76f285d2/nixos/modules/services/networking/pixiecore.nix#L15-L21

> It would be also great if you could update the wiki after that.

Yep. I also have another PR in the work that might be nice.

File: nixos/modules/misc/ids.nix [hunk 12]
```diff
@@ -346,6 +346,7 @@ in
       zoneminder = 314;
       paperless = 315;
       #mailman = 316;  # removed 2019-08-30
+      pixiecore = 317;
```
Comment (Mic92) [inline nixos/modules/misc/ids.nix]: I don't think we need a statically allocated uid/gid here.
The service itself has no state to maintain.

```suggestion
```

File: nixos/modules/misc/ids.nix [hunk 13]
```diff
@@ -647,6 +648,7 @@ in
...
```
Comment (Mic92) [inline nixos/modules/misc/ids.nix]:
```suggestion
```

File: nixos/modules/services/networking/pixiecore.nix [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+        description = "Additional command line arguments to pass to Pixiecore";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.groups.pixiecore.gid = config.ids.gids.pixiecore;
+    users.users.pixiecore = {
+      description = "Pixiecore daemon user";
+      uid = config.ids.uids.pixiecore;
```
Comment (Mic92) [inline nixos/modules/services/networking/pixiecore.nix]:
```suggestion
```

File: nixos/modules/services/networking/pixiecore.nix [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+      extraArguments = mkOption {
+        type = types.listOf types.str;
+        default = [];
+        description = "Additional command line arguments to pass to Pixiecore";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.groups.pixiecore.gid = config.ids.gids.pixiecore;
```
Comment (Mic92) [inline nixos/modules/services/networking/pixiecore.nix]:
```suggestion
    users.groups.pixiecore = {}
```

**Review by Mic92 — CHANGES_REQUESTED** [review]

**Comment by bbigras** [comment]
I made the changes. Thanks.

I'll squash the commits at the end. Or I think you guys can do it automatically while merging.

**Comment by bbigras** [comment]
What's that `grahamcofborg-eval` test failing?

File: nixos/modules/services/networking/pixiecore.nix [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+      extraArguments = mkOption {
+        type = types.listOf types.str;
+        default = [];
+        description = "Additional command line arguments to pass to Pixiecore";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.groups.pixiecore = {}
```
Comment (samueldr) [inline nixos/modules/services/networking/pixiecore.nix]: missing `;` here.

**Comment by samueldr** [comment]
The "Details" link has the trace from Nix.

image

**Comment by bbigras** [comment]
Thanks. I made the change and squashed the commits.

File: nixos/modules/services/networking/pixiecore.nix [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+      debug = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Log more things that aren't directly related to booting a recognized client";
+      };
+
+      dhcpNoBind = mkOption {
+        type = types.bool;
+        default = false;
```
Comment (Mic92) [inline nixos/modules/services/networking/pixiecore.nix]: This option does not work because pixiecore does not run as root.
I think this needs some conditional.

File: nixos/modules/services/networking/pixiecore.nix [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+      openFirewall = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Open ports (67, 69 UDP and 4011, 'port', 'statusPort' TCP) in the firewall for Pixiecore.
+        '';
+      };
+
+      mode = mkOption {
+        description = "Which mode to use";
```
Comment (Mic92) [inline nixos/modules/services/networking/pixiecore.nix]: How about using `boot` as a default mode?

File: nixos/modules/services/networking/pixiecore.nix (hunk 17 above)
Comment (Mic92) [reply nixos/modules/services/networking/pixiecore.nix]: Maybe there is a capability we can add so listening without binding works.

Comment (bbigras) [reply nixos/modules/services/networking/pixiecore.nix]: I'm not familiar with listening without binding nor capabilities. Any ideas @danderson ?

Comment (Mic92) [reply nixos/modules/services/networking/pixiecore.nix]: I just tested it, if you add `cap_net_raw`, when `dhcpNoBind` is set, it works.

```
  AmbientCapabilities = "cap_net_bind_service cap_net_raw";
```

File: nixos/modules/services/networking/pixiecore.nix [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+        User = "pixiecore";
+        Restart = "always";
+        AmbientCapabilities = "cap_net_bind_service";
+        ExecStart = 
+          let
+            argString = 
+              if cfg.mode == "boot"
+              then [ "boot" cfg.kernel ]
+                   ++ optional (cfg.initrd != "") [ cfg.initrd ]
+                   ++ optional (cfg.cmdLine != "") [ "--cmdline=${lib.escapeShellArg cfg.cmdLine}" ]
```
Comment (Mic92) [inline nixos/modules/services/networking/pixiecore.nix]:
```suggestion
                   ++ optional (cfg.initrd != "") cfg.initrd
                   ++ optional (cfg.cmdLine != "") "--cmdline=${lib.escapeShellArg cfg.cmdLine}"
```

**Comment by bbigras** [comment]
I made the changes. Thanks.

Should `cap_net_raw` be added only when `dhcpNoBind` is used?

**Comment by Mic92** [comment]
> I made the changes. Thanks.
> 
> Should `cap_net_raw` be added only when `dhcpNoBind` is used?

It is not needed otherwise. So it sounds like a good idea.

**Comment by Lassulus** [comment]
are both low low ports (67 and 69) used by dhcp?

**Comment by danderson** [comment]
Port 69 is TFTP. Pixiecore does a fairly elaborate chainloading operation (described in https://github.com/danderson/netboot/blob/master/pixiecore/README.booting.md) to ensure that it works with the widest possible variety of weird and broken firmwares.

TL;DR: a PXE-booting client hits port 67 for DHCP, then port 4011 (BINL, an MS proprietary fork of PXE but that's a de-facto standard), then port 69 (TFTP to download iPXE with HTTP support), then the HTTP port (to finally download boot instructions, kernel+initrd, and finish the netboot).

**Comment by bbigras** [comment]
I made the change.

Is it OK like this?

https://github.com/NixOS/nixpkgs/blob/b1ee6627a45a74bbd8392173b76a7194e604e7f3/nixos/modules/services/networking/pixiecore.nix#L111-L114

**Review by Profpatsch — APPROVED** [review]
LGTM from my side, if @Mic92 is okay with it.

**Comment by bbigras** [comment]
Thanks everyone! :heart: 

Can someone close https://github.com/NixOS/nixpkgs/pull/62113 and https://github.com/NixOS/nixpkgs/issues/49740 please?
